compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -257,30 +265,91 @@ impl<'f> LoopInvariantContext<'f> {
 
     /// Checks whether a `block` is control dependent on any blocks after
     /// the given loop's header.
-    fn is_control_dependent_post_pre_header(&mut self, loop_: &Loop, block: BasicBlockId) {
+    fn is_control_dependent_post_pre_header(
+        &mut self,
+        loop_: &Loop,
+        block: BasicBlockId,
+        all_loops: &[Loop],
+    ) {
+        // The block is already known to be in a control dependent nested loop
+        // Thus, we can avoid checking for control dependence again.
+        if self.nested_loop_control_dependent_blocks.contains(&block) {
+            self.current_block_control_dependent = true;
+            return;
+        }
+
+        // Find all blocks between the current block and the loop header, exclusive of the current block and loop header themselves.
         let all_predecessors = Loop::find_blocks_in_loop(loop_.header, block, &self.cfg).blocks;
+        let all_predecessors = all_predecessors
+            .into_iter()
+            .filter(|&predecessor| predecessor != block && predecessor != loop_.header)
+            .collect::<Vec<_>>();
 
         // Reset the current block control dependent flag, the check will set it to true if needed.
         // If we fail to reset it, a block may be inadvertently labelled
         // as control dependent thus preventing optimizations.
         self.current_block_control_dependent = false;
 
-        // Need to accurately determine whether the current block is dependent on any blocks between
+        // Now check whether the current block is dependent on any blocks between
         // the current block and the loop header, exclusive of the current block and loop header themselves
-        if all_predecessors
-            .into_iter()
-            .filter(|&predecessor| predecessor != block && predecessor != loop_.header)
-            .any(|predecessor| self.is_control_dependent(predecessor, block))
+        if all_predecessors.iter().any(|predecessor| self.is_control_dependent(*predecessor, block))
         {
             self.current_block_control_dependent = true;
+            return;
+        }
+
+        self.is_nested_loop_control_dependent(loop_, block, all_loops, all_predecessors);
+    }
+
+    /// Determines if the `block` is in a nested loop that is control dependent
+    /// on a block in the outer loop.
+    /// If this is the case, we block hoisting as control is not guaranteed.
+    /// If the block is not control dependent on the inner loop itself, it will be marked appropriately
+    /// when the inner loop is being processed later.
```

Comment:
```suggestion
    /// If the block is not control dependent on the inner loop itself, it will be marked appropriately
    /// when the inner loop is processed later.
```